README.md
A simple autocomplete widget for Flutter that shows suggestion with a delay.

## Features

- Delay for showing suggestions
- Easy to customize design
- Easy to use apis

<!--
## Demo

![open street map search and pick](https://user-images.githubusercontent.com/69592754/179368498-fe392cdb-c321-46e8-ac4d-6b816e0a3758.png)
-->

<!-- ## Help Maintenance

I've been maintaining quite many repos these days and burning out slowly. If you could help me cheer up, buying me a cup of coffee will make my life really happy and get much energy out of it.

<a href="https://www.buymeacoffee.com/RtrHv1C" target="_blank"><img src="https://www.buymeacoffee.com/assets/img/custom_images/purple_img.png" alt="Buy Me A Coffee" style="height: auto !important;width: auto !important;" ></a> -->

## Getting Started

Import the following package in your dart file

```dart
import 'package:delayed_autocomplete/delayed_autocomplete.dart';
```

## Then Usage

    DeyaledAutocomplete(
        // how much time you wanna wait before getting and showing suggestions
        delayinMilliseconds: 1000,
        hintText: "Search",
        borderColor: Colors.blue,
        // this is the widget that will be shown in the list
        itemWidget: (dynamic object) {
          String name = object as String;
          return Container(
            height: 50,
            child: Center(
              child: Text(name),
            ),
          );
        },
        toDo: (String suggestion) async {
          //you can call your api here
          return await _getSuggestions(suggestion);
          // this list's items must be of same type as the object you passed in the itemWidget
        },
      )

<!--
# Video Tutorial

Click on the image below to view a video tutorial. It will redirect you to a youtube video.

- Video 1

[![Click here to view the tutorial](https://img.youtube.com/vi/VHDlC8wC9FI/0.jpg)](https://www.youtube.com/watch?v=VHDlC8wC9FI)

- Video 2

[![Click here to view the tutorial](https://img.youtube.com/vi/kZRrH3UlxeU/0.jpg)](https://www.youtube.com/watch?v=kZRrH3UlxeU)
-->


CHANGELOG.md
## 0.0.4

- hintText added
- Asynchronous loading of data


test/delayedautocomplete_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:delayed_autocomplete/delayed_autocomplete.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
  });
}


lib/delayed_autocomplete.dart
library delayed_autocomplete;

import 'dart:async';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/src/widgets/container.dart';
import 'package:flutter/src/widgets/framework.dart';

class DeyaledAutocomplete extends StatefulWidget {
  final Future<List> Function(String suggestion) toDo;
  final Widget Function(dynamic object) itemWidget;
  final int delayinMilliseconds;
  final Color borderColor;
  final String hintText;
  const DeyaledAutocomplete(
      {super.key,
      required this.toDo,
      required this.itemWidget,
      this.borderColor = Colors.blue,
      this.delayinMilliseconds = 500,
      this.hintText = 'Search'});

  @override
  State<DeyaledAutocomplete> createState() => _DeyaledAutocompleteState();
}

class _DeyaledAutocompleteState extends State<DeyaledAutocomplete> {
  Timer? _debounce;
  List suggestions = [];
  bool isLoading = false;
  final TextEditingController _searchController = TextEditingController();
  final FocusNode _focusNode = FocusNode();

  @override
  Widget build(BuildContext context) {
    OutlineInputBorder inputBorder = OutlineInputBorder(
      borderRadius: BorderRadius.all(Radius.circular(15.0)),
      borderSide: BorderSide(
        color: widget.borderColor,
      ),
    );
    OutlineInputBorder inputFocusBorder = OutlineInputBorder(
      borderRadius: BorderRadius.all(Radius.circular(15.0)),
      borderSide: BorderSide(color: widget.borderColor, width: 3.0),
    );
    return SingleChildScrollView(
      child: Column(
        children: [
          TextFormField(
              controller: _searchController,
              focusNode: _focusNode,
              decoration: InputDecoration(
                  hintText: widget.hintText,
                  border: inputBorder,
                  focusedBorder: inputFocusBorder,
                  prefixIcon: Icon(
                    Icons.search,
                  )),
              onChanged: (String value) {
                if (_debounce?.isActive ?? false) _debounce?.cancel();
                isLoading = true;
                setState(() {});
                _debounce = Timer(const Duration(milliseconds: 2000), () async {
                  widget.toDo(_searchController.text).then((value) {
                    suggestions = value;
                    isLoading = false;
                    setState(() {});
                  });
                });
              }),
          SizedBox(
            height: 10,
          ),
          isLoading == false
              ? StatefulBuilder(builder: ((context, setState) {
                  return ListView.builder(
                      shrinkWrap: true,
                      // scrollDirection: Axis.vertical,
                      physics: const NeverScrollableScrollPhysics(),
                      itemCount: suggestions.length,
                      itemBuilder: (context, index) {
                        return widget.itemWidget(suggestions[index]);
                      });
                }))
              : Container(
                  height: MediaQuery.of(context).size.height - 200,
                  child: Center(
                    child: CupertinoActivityIndicator(),
                  ),
                ),
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:delayed_autocomplete/delayed_autocomplete.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  Future<List> _getSuggestions(String suggestion) async {
    await Future.delayed(Duration(seconds: 2));
    return ["Apple", "Banana", "Orange", "Pineapple", "Mango"];
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: DeyaledAutocomplete(
        // how much time you wanna wait before getting and showing suggestions
        delayinMilliseconds: 1000,
        hintText: "Search",
        borderColor: Colors.blue,
        // this is the widget that will be shown in the list
        itemWidget: (dynamic object) {
          String name = object as String;
          return Container(
            height: 50,
            child: Center(
              child: Text(name),
            ),
          );
        },
        toDo: (String suggestion) async {
          //you can call your api here
          return await _getSuggestions(suggestion);
          // this list's items must be of same type as the object you passed in the itemWidget
        },
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
